### PR TITLE
feat: query plan caching

### PR DESCRIPTION
**Pull request by aqrln**
## Summary

This PR implements **query plan caching** for Prisma Client, effectively eliminating the query compilation overhead that was introduced by the architecture shift in Prisma 7.

### Background

In Prisma 6, the query engine ran as a separate Rust process (or native library addon), so query compilation and execution happened on a separate thread and never blocked the JavaScript event loop. In Prisma 7, the query compiler runs as a Wasm module on the main JavaScript thread and execution happens in JavaScript itself. While this eliminated the IPC overhead and simplified the architecture, it meant that every query now synchronously blocks the event loop during compilation (typically 0.1-1ms per query). Under load, this compilation overhead adds up and contends with other work on the event loop.

### Key idea

In a typical application, query *shapes* are mostly static -- only the user-provided *values* change between calls. For example:

```ts
prisma.user.findUnique({ where: { email: 'alice@prisma.io' } })
prisma.user.findUnique({ where: { email: 'bob@prisma.io' } })
```

These two queries have the same structure and produce the same query plan. Only the email value differs. If we can extract the varying values and normalize the query into a canonical shape, we can use that shape as a cache key and skip recompilation entirely on subsequent calls.

### Solution

The implementation introduces **schema-aware query parameterization**: before sending a query to the Wasm compiler, the runtime replaces user-provided data values with typed placeholders (`%1`, `%2`, ...), producing a stable cache key. On a cache hit, the previously compiled query plan is reused and only the placeholder values are swapped in at execution time.

The parameterization is *schema-aware* rather than heuristic-based. A **ParamGraph** (parameterization graph) is built at generator time from the DMMF and embedded in the generated client. At runtime, the graph tells the parameterizer exactly which fields accept scalar values, enums, lists, or nested objects, and what types are expected, so it only parameterizes values it can safely replace.

### Architecture

```
Generator time:                    Runtime:

DMMF ──> ParamGraph ──> serialize   deserialize ──> ParamGraph
         (builder)      (binary)                    (runtime view)
                            │                            │
                            ▼                            ▼
                     Generated Client            Parameterizer
                                                      │
                                              ┌───────┴────────┐
                                              ▼                ▼
                                         Cache Key       Placeholder
                                        (normalized       Values
                                          query)        (extracted)
                                              │                │
                                              ▼                ▼
                                        LRU Cache ──> Query Plan ──> Executor
```

**Generator time:**
1. The DMMF is traversed to build a `ParamGraph` -- a compact directed graph describing which query arguments are parameterizable and what types they accept.
2. The graph is serialized into a hybrid format: a string table (JSON array) + structural data (base64url-encoded binary with varint encoding). This keeps the embedded schema small.

**Runtime (per query):**
1. The `ParamGraph` is deserialized once at `PrismaClient` construction.
2. For each query, the parameterizer walks the `JsonQuery` tree guided by the graph, replacing qualifying values with typed placeholders.
3. The parameterized query is serialized to JSON and used as an LRU cache key.
4. On a cache miss, the Wasm compiler compiles the query and the plan is stored.
5. On a cache hit (the common case), compilation is skipped entirely.
6. The executor receives the cached plan + the extracted placeholder values.

### Impact

- **~100% cache hit rate** in typical applications, since query shapes are mostly static.
- **Compilation cost drops from 0.1-1ms to 1-10µs** per query on cache hits (just the parameterization + cache lookup).
- Effectively eliminates the event loop contention issue introduced by the Wasm architecture.
- Both single queries and batched queries benefit from caching.

### New packages

- **`@prisma/param-graph`**: Runtime ParamGraph representation, binary serialization/deserialization, edge flags, and scalar masks.
- **`@prisma/param-graph-builder`**: Generator-time construction of the ParamGraph from DMMF. Includes iterative DMMF traversal (to avoid stack overflow on complex schemas) and conservative union type merging.

### Changes by area

**Query parameterization** (`packages/client/.../parameterization/`):
- `classify.ts` -- Value classification (null, primitive, tagged scalar, structural, array, object).
- `parameterize.ts` -- Core parameterization algorithm. Schema-guided tree walk with placeholder deduplication (identical values share placeholders so the query compiler can reason about value equality).

**Query plan cache** (`packages/client/.../query-plan-cache.ts`):
- LRU cache with separate maps for single and batch query plans (1000 entries per map by default).

**Client engine integration** (`ClientEngine.ts`):
- Wires parameterization + cache lookup into the single-query and batch-query request paths.
- Passes `placeholderValues` through the execution chain to the interpreter.

**Batch decompaction** (`packages/client-engine-runtime/src/batch.ts`):
- `convertCompactedRows` now resolves placeholders in compacted batch arguments before matching rows to results.

**Code generation** (`client-generator-js`, `client-generator-ts`):
- Both generators build and embed the serialized ParamGraph in the generated client configuration.

**JSON protocol** (`packages/json-protocol`):
- New `Placeholder`, `PlaceholderType`, and `PlaceholderTaggedValue` types for representing parameterized values in the JSON wire format.

### Test plan

- [x] Unit tests for ParamGraph serialization round-trips (including varint boundary values and optional field encoding)
- [x] Unit tests for parameterization across all value types: scalars, tagged values (DateTime, Decimal, BigInt, Bytes, Json), enums, arrays, nested objects, structural values (FieldRef, Enum tags)
- [x] Unit tests for filter operators (equals, gt, lt, gte, lte, contains, startsWith, endsWith, in, notIn, not)
- [x] Unit tests for placeholder naming determinism and deduplication
- [x] Unit tests for cache key consistency (same shape -> same key, different shape -> different key)
- [x] Unit tests for batch parameterization with cross-query placeholder uniqueness
- [x] Unit tests for LRU cache behavior (eviction, refresh on access, capacity limits)
- [x] Benchmarks for parameterization and caching overhead (CodSpeed integration)
- [x] Existing functional and integration test suites pass

### CODE REVIEW

File: packages/param-graph/src/serialization.test.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+    const serialized = serializeParamGraph(data)
+
+    // Decode base64url to check actual binary size
+    const binarySize = Buffer.from(serialized.graph, 'base64url').length
+
+    // With varint encoding, small indices (0-127) use 1 byte each
+    // The old compact format used 2 bytes per index plus padding
+    // This should be significantly smaller
+    expect(binarySize).toBeLessThan(80) // Empirically determined reasonable upper bound
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph/src/serialization.test.ts]: Ooof missed this one, I don't like this test, whether with a magic number or with a constant.

File: packages/param-graph/package.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  "license": "Apache-2.0",
+  "scripts": {
+    "dev": "DEV=true tsx helpers/build.ts",
+    "build": "tsx helpers/build.ts",
+    "prepublishOnly": "pnpm run build",
+    "test": "vitest run"
+  },
+  "devDependencies": {
+    "vitest": "^2.1.8"
+  },
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph/package.json]: 🤯

I don't think we should even have it in the dependencies at all here, the standard in the codebase is to use the vitest package installed in the root of the monorepo

File: packages/param-graph/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  RootEntryData,
+} from './param-graph'
+
+// Constants and utilities
+export type { EdgeFlagValue, ScalarMaskValue } from './param-graph'
+export { EdgeFlag, getScalarMask, hasFlag, ScalarMask, scalarTypeToMask } from './param-graph'
+
+// Serialization
+export type { SerializedParamGraph } from './serialization'
+export { deserializeParamGraph, serializeParamGraph } from './serialization'
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph/src/index.ts]: This is a package, not just a directory, the barrel file rule doesn't apply. The point about comments is valid.

File: packages/client-generator-js/src/TSClient/TSClient.ts [hunk 4]
```diff
@@ -47,14 +48,23 @@ export class TSClient implements Generable {
...
   }
 
+  private buildParamGraphConfig(): string {
+    const serialized = buildAndSerializeParamGraph(this.options.dmmf)
+    const stringsJson = JSON.stringify(JSON.stringify(serialized.strings))
+    return `config.parameterizationSchema = {
+  strings: JSON.parse(${stringsJson}),
+  graph: "${serialized.graph}"
+}`
+  }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client-generator-js/src/TSClient/TSClient.ts:58]: The graph value is base64url-encoded, which by definition only contains [A-Za-z0-9_-]. There's no escaping concern.

File: packages/client-generator-ts/src/TSClient/file-generators/ClassFile.ts [hunk 5]
```diff
@@ -60,12 +63,23 @@ function clientConfig(context: GenerateContext, options: TSClientOptions) {
...
 }
+
+function buildParameterizationSchema(dmmf: DMMF.Document): string {
+  const serialized = buildAndSerializeParamGraph(dmmf)
+  const stringsJson = JSON.stringify(JSON.stringify(serialized.strings))
+  return `config.parameterizationSchema = {
+  strings: JSON.parse(${stringsJson}),
+  graph: "${serialized.graph}"
+}`
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client-generator-ts/src/TSClient/file-generators/ClassFile.ts:85]: The graph value is base64url-encoded, which by definition only contains [A-Za-z0-9_-]. There's no escaping concern.

File: packages/client/src/__tests__/benchmarks/query-performance/qc-loader.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          const queryCompilerWasmFilePath = path.join(runtimeBase, `query_compiler_fast_bg.${provider}.wasm-base64.js`)
+          const wasmBase64: string = require(queryCompilerWasmFilePath).wasm
+          moduleBytes = Buffer.from(wasmBase64, 'base64')
+        }
+        return new WebAssembly.Module(moduleBytes)
+      },
+      importName: `./query_compiler_fast_bg.js`,
+    },
+  })
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/__tests__/benchmarks/query-performance/qc-loader.ts:38]: Test utility within the benchmark suite. A doc comment would add no value.

File: packages/client/src/runtime/core/engines/client/ClientEngine.ts [hunk 7]
```diff
@@ -493,13 +529,41 @@ export class ClientEngine implements Engine {
...
+    const batchPayload = getBatchRequestPayload(queries, transaction)
+    const request = JSON.stringify(batchPayload)
 
     const { executor, queryCompiler } = await this.#ensureStarted().catch((err) => {
-      throw this.#transformRequestError(err, JSON.stringify(request))
+      throw this.#transformRequestError(err, request)
     })
 
-    const batchResponse = this.#compileBatch(request, queryCompiler)
+    const hasRawQueries = firstModelName === undefined
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/ClientEngine.ts:534]: isRawQuery is a different check for a different context. `modelName === undefined` is correct here.

File: packages/client-engine-runtime/src/batch.ts [hunk 8]
```diff
@@ -19,11 +19,69 @@ export type CompactedBatchResponse = {
...
+    resolved[key] = value
+    if (isPlaceholder(value)) {
+      const placeholderName = getPlaceholderName(value)
+      if (placeholderName && placeholderName in placeholderValues) {
+        resolved[key] = placeholderValues[placeholderName] as {}
+      }
+    }
+  }
+  return resolved
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client-engine-runtime/src/batch.ts]: The arguments in a CompactedBatchResponse are not the original Prisma Client where clause shape -- they're the flattened column-level key-value pairs produced by the query compiler for row matching. The
keys field tells you which SQL result columns to compare against, and arguments contains one entry per original query with those same flat keys.

For a simple @id field:
```
keys: ["id"]
arguments: [{ id: <placeholder> }, { id: <placeholder> }, ...]
```

For a compound @@unique([email, name]), the query compiler still flattens to SQL-level columns:
```
keys: ["email", "name"]
arguments: [{ email: <placeholder>, name: <placeholder> }, ...]
```

It's never:
```
arguments: [{ email_name: { email: <placeholder>, name: <placeholder> } }]
```

That nested shape exists in the Prisma Client API (`where: { email_name: { email: "...", name: "..." } }`), but by the time the query compiler produces a compacted batch response, it's been resolved to flat
column keys. That's the whole point of compaction -- it maps back to SQL result rows by column name.

So the top-level-only resolution is correct by construction.

File: packages/param-graph-builder/src/dmmf-traverser.ts [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+        hasAnyEdge = true
+      }
+    }
+
+    if (hasAnyEdge) {
+      this.#builder.setOutputNodeEdges(nodeId, edges)
+    }
+
+    return nodeId
+  }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph-builder/src/dmmf-traverser.ts:418]: The deferred queue was introduced specifically for input types because the DMMF input type graph has combinatorial fan-out (each model generates dozens of filter/orderBy/create/update types, each referencing other models' types). Output types map 1:1 to models, so the recursion depth is just the depth of the model relation graph with cycles short-circuited by pre-allocation. Adding the deferred queue machinery to output types would add complexity with no practical benefit.

File: packages/param-graph-builder/src/param-graph-builder.ts [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  /**
+   * Sets edges on an input node.
+   */
+  setInputNodeEdges(nodeId: NodeId, edges: Record<number, InputEdgeData>): void {
+    if (Object.keys(edges).length > 0) {
+      this.#inputNodes[nodeId].edges = edges
+    }
+  }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph-builder/src/param-graph-builder.ts:65]: I think it will fail quite loudly

File: packages/param-graph/src/serialization.ts [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+      }
+    }
+
+    for (const key of this.#rootKeys) {
+      const root = this.#data.roots[key]
+      const keyIndex = this.#data.strings.indexOf(key)
+      size += varuintSize(keyIndex)
+      size += varuintSize(root.argsNodeId === undefined ? 0 : root.argsNodeId + 1)
+      size += varuintSize(root.outputNodeId === undefined ? 0 : root.outputNodeId + 1)
+    }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph/src/serialization.ts:240]: we care about deserialization speed, in serialization this is fine.

File: packages/client/src/runtime/core/engines/client/ClientEngine.ts [hunk 12]
```diff
@@ -135,6 +142,14 @@ export class ClientEngine implements Engine {
...
     this.datamodel = config.inlineSchema
     this.tracingHelper = config.tracingHelper
+    this.#queryPlanCache = new QueryPlanCache()
+    this.#queryPlanCacheEnabled = true
+    this.#paramGraph = ParamGraph.deserialize(config.parameterizationSchema, (enumName) => {
+      if (!Object.hasOwn(config.runtimeDataModel.enums, enumName)) {
+        return undefined
+      }
+      return config.runtimeDataModel.enums[enumName].values.map((v) => v.name)
+    })
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/ClientEngine.ts:149]: The enum lookup validates user-provided values against the Prisma schema names, not database names. Users pass Prisma names (ACTIVE, INACTIVE) in their queries, and the @Map directive only affects the query compiler's SQL generation layer. Using v.name here is correct

File: packages/client/src/runtime/core/engines/client/parameterization/parameterize.test.ts [hunk 13]
```diff
@@ -0,0 +1,687 @@
...
+import type { ParamGraphData } from '@prisma/param-graph'
+import { EdgeFlag, ParamGraph, ScalarMask } from '@prisma/param-graph'
+
+import { parameterizeBatch, parameterizeQuery } from './parameterize'
+
+function createEnumLookup(runtimeDataModel: RuntimeDataModel) {
+  return (enumName: string): readonly string[] | undefined => {
+    const enumDef = runtimeDataModel.enums[enumName]
+    return enumDef?.values.map((v) => v.name)
+  }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/parameterization/parameterize.test.ts:12]: responded above

File: packages/client/tests/functional/string-filters/prisma/_schema.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  datasource db {
+    provider = "${provider}"
+  }
+
+  model TestModel {
+    id    ${idForProvider(provider)}
+    value String
+  }
+  `
+})
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/functional/string-filters/prisma/_schema.ts:19]: _schema.ts is a special case

File: packages/client/src/runtime/core/engines/client/parameterization/parameterize.test.ts [hunk 15]
```diff
@@ -0,0 +1,687 @@
...
+          arguments: { where: dateValue },
+          selection: { $scalars: true },
+        },
+      }
+
+      const result = parameterizeQuery(query, graphWithDateTime)
+
+      // DateTime should be parameterized with decoded value
+      expect(result.placeholderValues['%1']).toBe('2024-01-01T00:00:00.000Z')
+    })
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/parameterization/parameterize.test.ts:277]: fixed

File: packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts [hunk 16]
```diff
@@ -309,6 +312,7 @@ function serializeArgumentsValue(
 
   if (ArrayBuffer.isView(jsValue)) {
     const { buffer, byteOffset, byteLength } = jsValue
+    // TODO(perf): get rid of this conversion
     return { $type: 'Bytes', value: Buffer.from(buffer, byteOffset, byteLength).toString('base64') }
   }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts:317]: That's obviously not what the comment is about (this ArrayBuffer view to Buffer conversion is zero cost due to the specific Buffer constructor used — it just becomes another view to the same ArrayBuffer). The comment is about the conversion to base64. No need to track it separately.

File: packages/param-graph/src/serialization.ts [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+      const keyIndex = this.#data.strings.indexOf(key)
+      if (keyIndex === -1) {
+        throw new Error(`Root key "${key}" not found in strings table`)
+      }
+
+      this.#writeVaruint(keyIndex)
+      this.#writeOptionalVaruint(root.argsNodeId)
+      this.#writeOptionalVaruint(root.outputNodeId)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph/src/serialization.ts:295]: I don't think it matters.

File: packages/client/tests/functional/string-filters/_matrix.ts [hunk 18]
```diff
@@ -0,0 +1,4 @@
+import { defineMatrix } from '../_utils/defineMatrix'
+import { allProviders } from '../_utils/providers'
+
+export default defineMatrix(() => [allProviders])
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/functional/string-filters/_matrix.ts:4]: This doesn't apply to _schema.ts, _matrix.ts etc.

File: packages/client/src/__tests__/benchmarks/query-performance/caching.bench.ts [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+let QueryCompilerClass: QueryCompilerConstructor
+let queryCompiler: QueryCompiler
+
+const BENCHMARK_DATAMODEL = fs.readFileSync(path.join(__dirname, 'schema.prisma'), 'utf-8')
+
+let paramGraph: ParamGraph
+
+type Param = { $type: 'Param'; value: string }
+type Parameterizable<T> = T | Param
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/__tests__/benchmarks/query-performance/caching.bench.ts:23]: It's used by the Parameterizable<T> definition and I don't think it should be inlined.

File: packages/client/src/runtime/core/engines/client/parameterization/parameterize.ts [hunk 20]
```diff
@@ -0,0 +1,557 @@
...
+
+    if (hasFlag(edge, EdgeFlag.ParamListScalar)) {
+      const allValid = items.every((item) => validateListElement(item, edge))
+      if (allValid && items.length > 0) {
+        const decodedItems = items.map((item) => decodeIfTagged(item))
+        const innerType = inferListElementType(items[0])
+        const type: PlaceholderType = { type: 'List', inner: innerType }
+        return this.#getOrCreatePlaceholder(decodedItems, type)
+      }
+    }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/parameterization/parameterize.ts:296]: fixed

File: packages/param-graph/package.json [hunk 21]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "@prisma/param-graph",
+  "version": "0.0.0",
+  "description": "This package is intended for Prisma's internal use",
+  "main": "dist/index.js",
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph/package.json:5]: good point but it follows the existing pattern

File: packages/param-graph/src/index.ts [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+  OutputNodeData,
+  ParamGraphData,
+  RootEntryData,
+} from './param-graph'
+
+export type { EdgeFlagValue, ScalarMaskValue } from './param-graph'
+export { EdgeFlag, getScalarMask, hasFlag, ScalarMask, scalarTypeToMask } from './param-graph'
+
+export type { SerializedParamGraph } from './serialization'
+export { deserializeParamGraph, serializeParamGraph } from './serialization'
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/param-graph/src/index.ts:14]: fixed

File: packages/client/src/runtime/core/engines/client/parameterization/parameterize-tests/batch.test.ts [hunk 23]
```diff
@@ -0,0 +1,414 @@
+import type { JsonBatchQuery } from '@prisma/json-protocol'
+
+import type { ParamGraph } from '@prisma/param-graph'
+
+import { parameterizeBatch } from '../parameterize'
+import { getParamGraph } from './test-fixtures'
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/parameterization/parameterize-tests/batch.test.ts:5]: fixed

File: packages/client/src/runtime/core/engines/client/parameterization/parameterize-tests/test-fixtures.ts [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    Status: {
+      values: [
+        { name: 'ACTIVE', dbName: null },
+        { name: 'INACTIVE', dbName: null },
+      ],
+      dbName: null,
+    },
+  },
+  types: {},
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/parameterization/parameterize-tests/test-fixtures.ts:71]: This rule says how these comments should be written if they are, not that they are almost required. This internal testing utility absolutely doesn't need a doc comment.

File: packages/client/src/runtime/core/engines/client/parameterization/parameterize.ts [hunk 25]
```diff
@@ -0,0 +1,557 @@
...
+    case 'Bytes':
+    case 'DateTime':
+    case 'Json':
+      return { type: tag }
+    case 'Decimal':
+      return { type: 'Float' }
+    default:
+      return undefined
+  }
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/src/runtime/core/engines/client/parameterization/parameterize.ts:474]: done

File: packages/client/src/runtime/core/engines/client/parameterization/parameterize.ts [hunk 26]
```diff
@@ -0,0 +1,588 @@
...
+/**
+ * Returns the wider of two placeholder types, following numeric promotion rules:
+ * Int -> BigInt -> Float. Non-numeric types must match exactly or produce Any.
+ */
+function widenType(a: PlaceholderType, b: PlaceholderType): PlaceholderType {
+  if (a.type === 'Any') return b
+  if (b.type === 'Any') return a
+  if (a.type === b.type) return a
+
+  const NUMERIC_WIDTH: Partial<Record<string, number>> = { Int: 0, BigInt: 1, Float: 2 }
```
Comment (jacek-prisma) [inline packages/client/src/runtime/core/engines/client/parameterization/parameterize.ts]:
```suggestion
  const NUMERIC_WIDTH: Partial<Record<PlaceholderType['type'], number>> = { Int: 0, BigInt: 1, Float: 2 }
```

File: packages/param-graph-builder/src/dmmf-traverser.ts [hunk 27]
```diff
@@ -0,0 +1,453 @@
...
+                  ot.type === inputType.type && ot.namespace === inputType.namespace && ot.isList === inputType.isList,
+              )
+            ) {
+              inputObjectTypes.push(inputType)
+            }
+            break
+          case 'fieldRefTypes':
+            break
+          default:
+            inputType.location satisfies never
```
Comment (jacek-prisma) [inline packages/param-graph-builder/src/dmmf-traverser.ts]: I think we should have a runtime assertion here, since the value comes from Rust and we currently do not have any static guarantee that it matches the typescript definition

**Review by jacek-prisma — APPROVED** [review]
LGTM, with 2 minor comments

**Review by jacek-prisma — APPROVED** [review]
